rust/wasmtime_bindings/src/memory_creator.rs

```diff
@@ -0,0 +1,97 @@
+use anyhow::{anyhow, Result};
+use wasmtime::LinearMemory;
+
+pub struct ScyllaLinearMemory {
+    ptr: *mut u8,
+    layout: Option<std::alloc::Layout>,
+    maximum_size: Option<usize>,
+}
+
+// The entire ScyllaLinearMemory is used only in a single thread,
+// because we're not sharing it between seastar shards
+unsafe impl Send for ScyllaLinearMemory {}
+unsafe impl Sync for ScyllaLinearMemory {}
+
+impl Drop for ScyllaLinearMemory {
+    fn drop(&mut self) {
+        if let Some(layout) = self.layout {
+            // previously allocated in grow_to()
+            unsafe { std::alloc::dealloc(self.ptr, layout) };
+        }
+    }
+}
+
+unsafe impl LinearMemory for ScyllaLinearMemory {
+    fn byte_size(&self) -> usize {
+        match self.layout {
+            Some(layout) => layout.size(),
+            None => 0,
+        }
+    }
+    fn maximum_byte_size(&self) -> Option<usize> {
+        self.maximum_size
+    }
+    fn grow_to(&mut self, new_size: usize) -> Result<()> {
+        if new_size == self.byte_size() {
+            return Ok(());
+        }
+        let max_size = self.maximum_size.unwrap_or(std::u32::MAX as usize);
+        assert!(new_size <= max_size);
+        if new_size == 0 {
+            // we know that current size is not 0, so we can deallocate it
+            unsafe { std::alloc::dealloc(self.ptr, self.layout.unwrap()) };
+            self.layout = None;
+            self.ptr = std::ptr::null_mut();
+            Ok(())
+        } else {
+            // new_size > 0, so we can use std::alloc to allocate a new memory
+            if let Ok(new_layout) = std::alloc::Layout::from_size_align(new_size, 64 * 1024) {
+                let new_ptr = match self.layout {
+                    Some(layout) => unsafe { std::alloc::realloc(self.ptr, layout, new_layout.size()) },
+                    None => unsafe { std::alloc::alloc(new_layout) },
```

Comment:
just to confirm - is this whole thing done simply to force wasm to use host malloc/free for allocations, so that it gets automatically handled by the Seastar allocator?

Reply:
Yes, otherwise the memory will be allocated with `mmap` that we don't override in seastar

Reply:
In this case, let's explicitly explain it in the comment, along with mentioning mmap. Technically, nothing guarantees that `std::alloc` doesn't switch to mmap one day

Reply:
(or, just use an unsafe malloc/free pair directly to be sure)

Reply:
Oh wait, it's an unsafe block anyway. In this case I strongly suggest you just forward-declare malloc, realloc, free, whatever's needed, and use them directly. Then we're 100% sure there won't be any mmap sneaking in.

And in case we have strict layout requirements, like a 64k page for wasm, perhaps instead of malloc we should just go with aligned_alloc (https://linux.die.net/man/3/aligned_alloc)

Reply:
Ref: https://github.com/rust-lang/rust/blob/c3605f8c8020dbbe8f0d1961c7b33c4c4b78ad0d/library/std/src/sys/unix/alloc.rs#L6-L24

with this implementation it uses aligned_alloc (or malloc for small allocations), which we properly override in Seastar indeed, and it will most likely stay that way due to the 64k alignment, because mmap doesn't exactly let you choose a multi-page alignment anyway. But in theory on on certain hugepage configurations it may be beneficial to use mmap over aligned_alloc, so in my opinion we should be prepared for that

Reply:
Ok, I used  `aligned_alloc` and `free` in the new rebase. I've also added a comment about mmap.
I removed the `std::alloc::realloc` because it should probably not be used here anyway, considering it doesn't keep the alignment